### PR TITLE
Add QED particle creaction routines

### PR DESCRIPTION
**Pull request by lucafedeli88**
This pull request implements QED particle creation routines (pair generation via the Breit-Wheeler process and photon emission). 

#### Main changes:
- The `QED` folder in `Source` has been moved to `Source/Particles/ElementaryProcess/QEDInternals` 

- `Source/Particles/ElementaryProcess/QEDPhotonEmission.H` provides functors to be used with the new particle creation routines to implement photon emission.

- `Source/Particles/ElementaryProcess/QEDPairGeneration.H` provides functors to be used with the new particle creation routines to implement pair generation.

- In `MultiParticleContainer.H` a new version of `MFItInfoCheckTiling` has been implemented, in order to take into account multiple destination species.

- In  `MultiParticleContainer.cpp`, the  `mapSpeciesProduct ` function has been modified to pair source and product species for the QED processes

- In `MultiParticleContainer.cpp` the two functions `doQedBreitWheeler` and `doQedQuantumSync` have been implemented. It is worth to remark that `doQedQuantumSync` calls `cleanLowEnergyPhotons`. The purpose of this function is to remove right after creation low energy photons which cannot generate pairs. This is achieved by setting to -1 the id of the low energy photons. 

- In `FilterCopyTransform.H ` some functions have been modified to better suit the case of two destination species.

- QED tests have been updated. **Warning** : QED tests are incomplete. For Breit Wheeler process only optical depth initialization and evolution are checked. For photon emission only optical depth initialization is checked.

- In `PhysicalParticleContainer` an `enum  PhysicalParticleType` has been added. The possible values are `photon`, `electron`, `positron`, `other`. Constructors of `PhysicalParticleContainer` and `PhotonParticleContainer` set a private `PhysicalParticleType` variable. This variable is used in methods `AmIAPhoton`, `AmIAPositron` and `AmIAnElectron`. This strategy has been adopted to fix an issue related to virtual function calls in constructors found by `lgtm`.

### CODE REVIEW

File: Source/Particles/MultiParticleContainer.H
Comment (ax3l) [inline Source/Particles/MultiParticleContainer.H]: alignment slightly off :)

File: Source/Particles/MultiParticleContainer.H
Comment (ax3l) [inline Source/Particles/MultiParticleContainer.H]:
```suggestion
```

File: Source/Particles/MultiParticleContainer.cpp
Comment (ax3l) [inline Source/Particles/MultiParticleContainer.cpp]:
```suggestion
                constexpr auto energy_threshold = amrex::ParticleReal(
```

File: Source/Particles/PhysicalParticleContainer.H
Comment (ax3l) [inline Source/Particles/PhysicalParticleContainer.H]:
```suggestion
    bool AmIAPhoton () override {return (m_my_type == PhysicalParticleType::photon);};
```

File: Source/Particles/PhysicalParticleContainer.H
Comment (ax3l) [inline Source/Particles/PhysicalParticleContainer.H]:
```suggestion
    bool AmIAnElectron () override {return (m_my_type == PhysicalParticleType::electron);};
```

File: Source/Particles/PhysicalParticleContainer.H
Comment (ax3l) [inline Source/Particles/PhysicalParticleContainer.H]:
```suggestion
    bool AmIAPositron () override {return (m_my_type == PhysicalParticleType::positron);};
```

File: Source/Particles/PhysicalParticleContainer.H
Comment (ax3l) [inline Source/Particles/PhysicalParticleContainer.H]:
```suggestion
```

File: Source/Particles/PhysicalParticleContainer.H
Comment (ax3l) [inline Source/Particles/PhysicalParticleContainer.H]:
```suggestion
```

**Review by ax3l — COMMENTED** [review]
Reads nice! Great API doc strings!

**Comment by MaxThevenet** [comment]
Just a small note: there could be conflicts to solve with PR https://github.com/ECP-WarpX/WarpX/pull/738.

**Review by MaxThevenet — COMMENTED** [review]
Thanks for this PR, looking forward to have it working :).

This is a partial review, but here are the main points:
- For the species type (electron, positron, etc.), I think it would make sense to have these as input parameters instead of guessing from mass and charge. I can make a PR for that.
- As discussed before, I think it is confusing to use `tau` for two different quantities, I'll make a PR to use two different names.
- I did not get into the details, but it seems to me that the tau initialization (`enable_opt_depth_for_targets` etc.) could be massively simplified by the `InitializationPolicy` that @atmyers introduced in PR https://github.com/ECP-WarpX/WarpX/pull/598. To my understanding, you can provide an initialization policy to particle quantities, and SmartCopy will apply this policy when doing the initialization, if the attribute exists in `dst`. @atmyers is that correct? I think things are made more complicated here because we use the name `tau` for two different things, which should be fixed, as mentioned above. @lucafedeli88 do you think you could use the `InitializationPolicy`?

Update:
One is done here: https://github.com/ECP-WarpX/WarpX/pull/742

File: Source/Particles/MultiParticleContainer.H
Comment (lucafedeli88) [reply Source/Particles/MultiParticleContainer.H]: ah, you are right!

File: Source/Particles/ElementaryProcess/QEDPairGeneration.H
Comment (lucafedeli88) [reply Source/Particles/ElementaryProcess/QEDPairGeneration.H]: you are right... I should do that

File: Source/Particles/ElementaryProcess/QEDPhotonEmission.H
Comment (lucafedeli88) [reply Source/Particles/ElementaryProcess/QEDPhotonEmission.H]: Sorry, I forgot to add this!

File: Source/Particles/ElementaryProcess/QEDPhotonEmission.H
Comment (lucafedeli88) [reply Source/Particles/ElementaryProcess/QEDPhotonEmission.H]: Thanks for the advice!

**Review by lucafedeli88 — COMMENTED** [review]
> Just a small note: there could be conflicts to solve with PR #738.

Thanks for pointing this out... I should have replaced all the occurrences of `BL_PROFILE` with `WARPX_PROFILE`

**Comment by lucafedeli88** [comment]
>     * For the species type (electron, positron, etc.), I think it would make sense to have these as input parameters instead of guessing from mass and charge. I can make a PR for that.

I totally agree! Thanks for https://github.com/ECP-WarpX/WarpX/pull/746 , as soon as it is merged I will make the necessary modifications. 

>     * As discussed before, I think it is confusing to use `tau` for two different quantities, I'll make a PR to use two different names.

I agree. Thanks for https://github.com/ECP-WarpX/WarpX/pull/742 ! As above, as soon as it is merged I will adapt this PR.
 
>     * I did not get into the details, but it seems to me that the tau initialization (`enable_opt_depth_for_targets` etc.) could be massively simplified by the `InitializationPolicy` that @atmyers introduced in PR #598. To my understanding, you can provide an initialization policy to particle quantities, and SmartCopy will apply this policy when doing the initialization, if the attribute exists in `dst`. @atmyers is that correct? I think things are made more complicated here because we use the name `tau` for two different things, which should be fixed, as mentioned above. @lucafedeli88 do you think you could use the `InitializationPolicy`?

That is correct. `tau` initialization is just a random number extraction from an exponential distribution. Actually, the same functor could be used for both electrons/positrons and photons. However, I prefer to keep them separate in case of future changes... As soon as  https://github.com/ECP-WarpX/WarpX/pull/742 is merged I will address this issue. As you mentioned this should make some methods much simpler...

File: Source/Particles/MultiParticleContainer.cpp
Comment (lucafedeli88) [reply Source/Particles/MultiParticleContainer.cpp]: we can try to make them `const` and see what happens...

Comment (lucafedeli88) [reply Source/Particles/MultiParticleContainer.cpp]: apparently, some of them can be `const`...

File: Source/Particles/ElementaryProcess/QEDPhotonEmission.H
Comment (lucafedeli88) [reply Source/Particles/ElementaryProcess/QEDPhotonEmission.H]: Actually optical depth initialization functors don't have member variables, so I guess that I can simply pass-by-copy here...

**Comment by lucafedeli88** [comment]
Thanks a lot for your reviews! I should have addressed all the points that you've raised. However, I have to wait for a couple of other PRs to be merged before performing some of the changes that you have suggested...

**Comment by MaxThevenet** [comment]
Is this still WIP?

File: Source/Particles/MultiParticleContainer.H
Comment (MaxThevenet) [inline Source/Particles/MultiParticleContainer.H]:
```suggestion
#   include "Particles/ElementaryProcess/QEDInternals/QedChiFunctions.H"
#   include "Particles/ElementaryProcess/QEDInternals/BreitWheelerEngineWrapper.H"
#   include "Particles/ElementaryProcess/QEDInternals/QuantumSyncEngineWrapper.H"
```
Now we add the paths to include statements.

File: Source/Particles/MultiParticleContainer.cpp
Comment (MaxThevenet) [inline Source/Particles/MultiParticleContainer.cpp]:
```suggestion
    WARPX_PROFILE("MPC::doQedBreitWheeler");
```

File: Source/Particles/MultiParticleContainer.cpp
Comment (MaxThevenet) [inline Source/Particles/MultiParticleContainer.cpp]:
```suggestion
                const auto& src_tile = pc_source->ParticlesAt(lev, mfi);
```
It may be worth trying putting a bunch of `const` here and there.

File: Source/Particles/MultiParticleContainer.cpp
Comment (MaxThevenet) [inline Source/Particles/MultiParticleContainer.cpp]:
```suggestion
                constexpr auto energy_threshold = static_cast<amrex::ParticleReal>(
                     2.0 * PhysConst::m_e * PhysConst::c * PhysConst::c);
```
?

File: Source/Particles/PhotonParticleContainer.cpp
Comment (MaxThevenet) [inline Source/Particles/PhotonParticleContainer.cpp]: Could use `WarpXParticleContainer::physical_species` directly? It should be meant for that.

File: Source/Particles/PhysicalParticleContainer.H
Comment (MaxThevenet) [inline Source/Particles/PhysicalParticleContainer.H]:
```suggestion
    #include "Particles/ElementaryProcess/QEDInternals/QEDPairGeneration.H"
    #include "Particles/ElementaryProcess/QEDInternals/QEDPhotonEmission.H"
```

**Review by MaxThevenet — CHANGES_REQUESTED** [review]
Looks great overall, thanks! I just have minor comments, see below.

File: Source/Evolve/WarpXEvolve.cpp
Comment (lucafedeli88) [reply Source/Evolve/WarpXEvolve.cpp]: sure!

File: Source/Particles/ElementaryProcess/QEDPairGeneration.H
Comment (lucafedeli88) [reply Source/Particles/ElementaryProcess/QEDPairGeneration.H]: Sure! I've added `_rt` suffix also in other places in `QEDPairGeneration.H` and `QEDPhotonEmission.H`. `const auto` is now `const amrex::ParticleReal`, in order to be explicit...

File: Source/Particles/MultiParticleContainer.cpp
Comment (lucafedeli88) [reply Source/Particles/MultiParticleContainer.cpp]: Unfortunately it seems that `src_tile` and `dst_tile` cannot be `const` due to the fact that they are passed to `filterCopyTransformParticles`. However, I've added a bunch of `const` as requested!

File: Source/Particles/MultiParticleContainer.cpp
Comment (lucafedeli88) [reply Source/Particles/MultiParticleContainer.cpp]: ah sure!

File: Source/Particles/PhysicalParticleContainer.H
Comment (lucafedeli88) [reply Source/Particles/PhysicalParticleContainer.H]: The path is actually just `Particles/ElementaryProcess/` , so I cannot commit directly your suggestion. However I have implemented the change.

File: Source/Particles/PhotonParticleContainer.cpp
Comment (lucafedeli88) [reply Source/Particles/PhotonParticleContainer.cpp]: ah, yes! sure...

Comment (lucafedeli88) [reply Source/Particles/PhotonParticleContainer.cpp]: I've also replaced all the `AmISomething` functions with a unique: 
```
    template<PhysicalSpecies PhysSpec>
    bool AmIA () const noexcept {return (physical_species == PhysSpec);}
```
in `WarpXParticleContainer.H`

File: Source/Particles/MultiParticleContainer.H
Comment (lucafedeli88) [reply Source/Particles/MultiParticleContainer.H]: Ok! I will comply with this policy in the future... I could not commit directly your suggestion because only `QedChiFunctions.H` is in `QEDInternals`, but I have implemented it.

File: Source/Particles/ElementaryProcess/QEDPhotonEmission.H
Comment (lucafedeli88) [reply Source/Particles/ElementaryProcess/QEDPhotonEmission.H]: If I have understood correctly, `InitializationPolicy` specifies how the components of the product species should be initialized. However, here we need to reset the optical depth component of the source species. is there a way to use `InitializationPolicy`  also for source species?

**Comment by lucafedeli88** [comment]
I've also updated the documentation and QED-related examples

File: Examples/Modules/qed/breit_wheeler/inputs_2d_tau_init
Comment (lucafedeli88) [reply Examples/Modules/qed/breit_wheeler/inputs_2d_tau_init]: I really like what you propose (including not allowing to create a species with just `photons.qed_breit_wheeler_ele_product_species = ele_bw`)!
Probably the easiest way to achieve this could be to allow for default values for most of the parameters of a species (except `mass` *and* `charge` or `species_type`).  The default value for `profile` could be something like `empty` or `do_not_generate`, so that the code could know that methods for particle initialization should not be called. 
Would something like that be OK? In case, I could work on a dedicated PR...

**Comment by lucafedeli88** [comment]
Sorry, some tests are failing... I am trying to understand why...

**Comment by lucafedeli88** [comment]
Ok... issues have been fixed. There was an error related to 1/m , due to the fact that now m=0 for photons.

File: Source/Particles/MultiParticleContainer.cpp
Comment (MaxThevenet) [inline Source/Particles/MultiParticleContainer.cpp]: Does this piece of code really belong here? Maybe we could put it in a new function
```c++
#ifdef WARPX_QED
MultiParticleContainer::CheckQEDSpecies(){
    ...
}
#endif
```
that we would call after `mapSpeciesProduct`, what do you think?

File: Source/Particles/MultiParticleContainer.cpp
Comment (MaxThevenet) [inline Source/Particles/MultiParticleContainer.cpp]: :) Thanks, sounds good.
```suggestion
                auto& src_tile = pc_source->ParticlesAt(lev, mfi);
```
^ Was this space here on purpose?

File: Source/Particles/MultiParticleContainer.cpp
Comment (MaxThevenet) [inline Source/Particles/MultiParticleContainer.cpp]:
```suggestion
```
Could we remove the three lines here? (I am not sure the Github suggestion is correct here, it might be trying to remove only one line, dunno why).

File: Source/Particles/MultiParticleContainer.cpp
Comment (MaxThevenet) [inline Source/Particles/MultiParticleContainer.cpp]:
```suggestion
        const auto Filter   = phys_pc_ptr->getPhotonEmissionFilterFunc();
        const auto CopyPhot = copy_factory_phot.getSmartCopy();
```

File: Source/Particles/MultiParticleContainer.cpp
Comment (MaxThevenet) [inline Source/Particles/MultiParticleContainer.cpp]: Should this be an input parameter?

File: Source/Particles/ElementaryProcess/QEDPhotonEmission.H
Comment (MaxThevenet) [reply Source/Particles/ElementaryProcess/QEDPhotonEmission.H]: Oh no, sorry, you're absolutely right, that looks perfect.

File: Source/Particles/ElementaryProcess/QEDPhotonEmission.H
Comment (MaxThevenet) [inline Source/Particles/ElementaryProcess/QEDPhotonEmission.H]:
```suggestion
    PTile& ptile,
    const int old_size, const int num_added,
    const amrex::ParticleReal t_energy_threshold)
```
or am I missing some indentation conventions?

File: Source/Particles/ElementaryProcess/QEDPhotonEmission.H
Comment (MaxThevenet) [inline Source/Particles/ElementaryProcess/QEDPhotonEmission.H]: Could you comment on what you are doing with the physical quantities here? `t_energy_threshold` is a physical energy, correct? So `energy_threshold` is a velocity? Maybe some comments in the code would help :).

**Review by MaxThevenet — COMMENTED** [review]
Looks good! Just have minor comments, we should be able to merge after minor changes :).

File: Examples/Modules/qed/breit_wheeler/inputs_2d_tau_init
Comment (MaxThevenet) [reply Examples/Modules/qed/breit_wheeler/inputs_2d_tau_init]: Yes, that sounds good! You are most welcome to look into it.

File: Source/Particles/PhotonParticleContainer.cpp
Comment (MaxThevenet) [reply Source/Particles/PhotonParticleContainer.cpp]: That look nice.

File: Source/Particles/MultiParticleContainer.cpp
Comment (lucafedeli88) [reply Source/Particles/MultiParticleContainer.cpp]: yes, we could make it an input parameter... I will do that

File: Source/Particles/ElementaryProcess/QEDPhotonEmission.H
Comment (lucafedeli88) [reply Source/Particles/ElementaryProcess/QEDPhotonEmission.H]: oh sorry! You are right!

File: Source/Particles/ElementaryProcess/QEDPhotonEmission.H
Comment (lucafedeli88) [reply Source/Particles/ElementaryProcess/QEDPhotonEmission.H]: I have re-organized the calculation, so that it should appear more transparent now...

File: Source/Particles/MultiParticleContainer.cpp
Comment (lucafedeli88) [reply Source/Particles/MultiParticleContainer.cpp]: I have created a new `CheckQEDProductSpecies` function as suggested, which now includes all the checks (except the test on the existence of the product species). For consistency I have created an analogous function for the Ionization process. In the future we might decide to ask the user to mark as `electrons` the product species and check for that...

File: Source/Particles/MultiParticleContainer.cpp
Comment (lucafedeli88) [reply Source/Particles/MultiParticleContainer.cpp]: I should have implemented the change. The energy threshold for photon creation is now an
optional input parameter. I have updated the documentation accordingly.

File: Source/Particles/MultiParticleContainer.cpp
Comment (MaxThevenet) [inline Source/Particles/MultiParticleContainer.cpp]: Some merge markers have survived here.

**Review by MaxThevenet — COMMENTED** [review]
Looks good, but there seems to be some merge conflict marks in one place.

**Review by MaxThevenet — APPROVED** [review]

**Comment by MaxThevenet** [comment]
:clap:

**Pull request merged**
